### PR TITLE
migrate async, poll functions, and samples from assistant to agent

### PR DESCRIPTION
**Pull request by howieleung**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/ai/azure-ai-client/samples/agents/sample_agents_file_search.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+from azure.ai.client.models._patch import FileSearchTool, ToolSet
+from azure.identity import DefaultAzureCredential
+
+# Set logging level
+logging.basicConfig(level=logging.INFO)
+
+# Create an Azure AI Client from a connection string, copied from your AI Studio project.
+# At the moment, it should be in the format "<HostName>;<AzureSubscriptionId>;<ResourceGroup>;<HubName>"
+# Customer needs to login to Azure subscription via Azure CLI and set the environment variables
+
```
Comment (dargilco) [inline sdk/ai/azure-ai-client/samples/agents/sample_agents_file_search.py:17]: Please use this "standard" header as you see in inference & endpoints samples. Put it below the copyright note.

"""
FILE: sample_get_chat_completions_client.py
DESCRIPTION:
    Given an AzureAIClient, this sample demonstrates how to get an authenticated 
    async ChatCompletionsClient from the azure.ai.inference package.
USAGE:
    python sample_get_chat_completions_client.py
    Before running the sample:
    pip install azure.ai.client azure-identity
    Set this environment variables with your own values:
    AI_CLIENT_CONNECTION_STRING - the Azure AI Project connection string, as found in your AI Studio Project.
"""

File: sdk/ai/azure-ai-client/samples/agents/async_samples/sample_agents_basics_async.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    logging_enable=True, # Optional. Remove this line if you don't want to show how to enable logging
+)
+"""
+async def main():
+    async with ai_client:
+        agent = await ai_client.agents.create_agent(
+            model="gpt-4-1106-preview", name="my-assistant", instructions="You are helpful assistant"
+        ) 
+        logging.info(f"Created agent, agent ID: {agent.id}")
+
```
Comment (dargilco) [inline sdk/ai/azure-ai-client/samples/agents/async_samples/sample_agents_basics_async.py:41]: Please just use "print" statements instead of "logging" to align with other samples. Are you using "logging" because you want to redirect logs to something other than the console?

**Review by dargilco — APPROVED** [review]

File: sdk/ai/azure-ai-client/azure/ai/client/operations/_patch.py [hunk 3]
```diff
@@ -1127,6 +1127,231 @@ def upload_file(
...
+    def upload_file_and_poll(self, body: JSON, sleep_interval: float = 1, **kwargs: Any) -> _models.OpenAIFile:
+        """Uploads a file for use by other operations.
+
+        :param body: Required.
+        :type body: JSON
+        :keyword sleep_interval: Time to wait before polling for the status of the uploaded file. Default value
+         is 1.
+        :paramtype sleep_interval: float
+        :return: OpenAIFile. The OpenAIFile is compatible with MutableMapping
+        :rtype: ~azure.ai.assistants.models.OpenAIFile
```
Comment (jhakulin) [inline sdk/ai/azure-ai-client/azure/ai/client/operations/_patch.py:1140]: please change rtype
:rtype: ~azure.ai.assistants.models.OpenAIFile

to

:rtype: ~azure.ai.client.models.OpenAIFile

In all places where you have azure.ai.assistants.models.OpenAIFile

File: sdk/ai/azure-ai-client/samples/agents/async_samples/sample_agents_basics_async.py (hunk 2 above)
Comment (jhakulin) [reply sdk/ai/azure-ai-client/samples/agents/async_samples/sample_agents_basics_async.py:41]: We used logging in my other branch.

That was selected instead of prints mainly I used them more, you can adjust filters etc.
We are fine to use prints also if that is the "standard" for samples

File: sdk/ai/azure-ai-client/samples/agents/sample_agents_basics.py [hunk 4]
```diff
@@ -29,32 +32,24 @@
     logging_enable=True, # Optional. Remove this line if you don't want to show how to enable logging
 )
```
Comment (jhakulin) [inline sdk/ai/azure-ai-client/samples/agents/sample_agents_basics.py:33]: Please do not change the existing samples, was there a reason?

File: sdk/ai/azure-ai-client/samples/agents/sample_agents_basics.py [hunk 5]
```diff
@@ -29,32 +32,24 @@
...
-run = ai_client.agents.create_run(thread_id=thread.id, assistant_id=agent.id)
-print("Created run, run ID", run.id)
+with ai_client:
+    agent = ai_client.agents.create_agent(
+        model="gpt-4-1106-preview", name="my-assistant", instructions="You are helpful assistant"
+    )
+    logging.info(f"Created agent, agent ID: {agent.id}")
 
-# poll the run as long as run status is queued or in progress
-while run.status in ["queued", "in_progress", "requires_action"]:
```
Comment (jhakulin) [inline sdk/ai/azure-ai-client/samples/agents/sample_agents_basics.py:47]: The while loop in basic sample is intentionally shown, the sample purpose is to show what is the lowest level api operation without convenience methods and which follows closest to openai api

**Review by jhakulin — CHANGES_REQUESTED** [review]
Please rename azure.ai.assistants in patch files to azure.ai.client
Please revert changes in existing samples

File: sdk/ai/azure-ai-client/samples/agents/async_samples/sample_agents_basics_async.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+"""
+ai_client = AzureAIClient(
+    credential=DefaultAzureCredential(),
+    host_name=os.environ["AI_CLIENT_HOST_NAME"],
+    subscription_id=os.environ["AI_CLIENT_SUBSCRIPTION_ID"],
+    resource_group_name=os.environ["AI_CLIENT_RESOURCE_GROUP_NAME"],
+    workspace_name=os.environ["AI_CLIENT_WORKSPACE_NAME"],
+    logging_enable=True, # Optional. Remove this line if you don't want to show how to enable logging
+)
+"""
```
[the remark this answers is not in the document]
Comment (howieleung) [reply sdk/ai/azure-ai-client/samples/agents/async_samples/sample_agents_basics_async.py:34]: It requires indentation making it look bad.

File: sdk/ai/azure-ai-client/samples/agents/sample_agents_file_search.py (hunk 1 above)
Comment (howieleung) [reply sdk/ai/azure-ai-client/samples/agents/sample_agents_file_search.py:38]: Done

File: sdk/ai/azure-ai-client/samples/agents/async_samples/sample_agents_basics_async.py (hunk 2 above)
Comment (howieleung) [reply sdk/ai/azure-ai-client/samples/agents/async_samples/sample_agents_basics_async.py:59]: Done

File: sdk/ai/azure-ai-client/azure/ai/client/operations/_patch.py (hunk 3 above)
Comment (howieleung) [reply sdk/ai/azure-ai-client/azure/ai/client/operations/_patch.py]: Done

File: sdk/ai/azure-ai-client/samples/agents/sample_agents_basics.py (hunk 4 above)
Comment (howieleung) [reply sdk/ai/azure-ai-client/samples/agents/sample_agents_basics.py:48]: Are you talking about the FILE, DESCRIPTION and USAGE?   I think Darron didn't include it in his first agent basic sample, so I follow the pattern.   But he just mentioned to add them.   So now I have them.

File: sdk/ai/azure-ai-client/samples/agents/sample_agents_basics.py (hunk 5 above)
Comment (howieleung) [reply sdk/ai/azure-ai-client/samples/agents/sample_agents_basics.py:47]: Done

File: sdk/ai/azure-ai-client/samples/agents/async_samples/sample_agents_basics_async.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dargilco) [reply sdk/ai/azure-ai-client/samples/agents/async_samples/sample_agents_basics_async.py:52]: Yes, it requires indentation. That's the nature of Python. Why does that look bad?

A sample function should be self-contained. It's weird to me that the client is defined outside the function.

[the remark this answers is not in the document]
Comment (howieleung) [reply sdk/ai/azure-ai-client/samples/agents/async_samples/sample_agents_basics_async.py]: I just pushed it.   Please take a look.   Samples with event handler really looks worse.   I prefer ai_client and connection string declared early.   I would prefer to rollback this change.

File: sdk/ai/azure-ai-client/samples/agents/sample_agents_basics.py [hunk 7]
```diff
@@ -29,32 +47,31 @@
...
-print("Created run, run ID", run.id)
+    message = ai_client.agents.create_message(thread_id=thread.id, role="user", content="Hello, tell me a joke")
+    print(f"Created message, message ID: {message.id}")
 
-# poll the run as long as run status is queued or in progress
-while run.status in ["queued", "in_progress", "requires_action"]:
-    # wait for a second
-    time.sleep(1)
-    run = ai_client.agents.get_run(thread_id=thread.id, run_id=run.id)
+    run = ai_client.agents.create_and_process_run(thread_id=thread.id, assistant_id=agent.id, sleep_interval=4)
```
Comment (jhakulin) [inline sdk/ai/azure-ai-client/samples/agents/sample_agents_basics.py:63]: please use create_run

**Review by jhakulin — APPROVED** [review]

File: sdk/ai/azure-ai-client/samples/agents/sample_agents_basics.py (hunk 4 above)
Comment (howieleung) [reply sdk/ai/azure-ai-client/samples/agents/sample_agents_basics.py:48]: According to our call, I have resolved this.

File: sdk/ai/azure-ai-client/samples/agents/sample_agents_stream_eventhandler_with_toolset.py [hunk 8]
```diff
@@ -61,8 +62,8 @@
 
 class MyEventHandler(AgentEventHandler):
 
-    def __init__(self, client: Agent):
-        self._client = client
+    def __init__(self, agents: AgentsOperations):
```
Comment (jhakulin) [inline sdk/ai/azure-ai-client/samples/agents/sample_agents_stream_eventhandler_with_toolset.py:65]: what is the reason for this change?

Comment (howieleung) [reply sdk/ai/azure-ai-client/samples/agents/sample_agents_stream_eventhandler_with_toolset.py:65]: Oh.   I found that this sample doesn't work after migration.   The function, submit_tool_outputs_to_run is in the AgentOperations.   Not in the Agent object.

Comment (jhakulin) [reply sdk/ai/azure-ai-client/samples/agents/sample_agents_stream_eventhandler_with_toolset.py:65]: ok, thanks!
